staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding_test.go

```diff
@@ -0,0 +1,304 @@
+		},
+		{
+			"deckSize: 128 handSize: 6",
+			128,
+			6,
+			func(int32) {},
+			true,
+		},
+		{
+			"deckSize: 1024 handSize: 6",
+			1024,
+			6,
+			func(int32) {},
+			true,
+		},
+		{
+			"deckSize: 512 handSize: 8",
+			512,
+			8,
+			func(int32) {},
+			false,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			if (ShuffleAndDealWithValidation(rand.Uint64(), test.deckSize, test.handSize, test.pick) == nil) != test.validated {
+				t.Errorf("test case %s fails", test.name)
+				return
+			}
+		})
+	}
+}
+
+func BenchmarkShuffleAndDeal(b *testing.B) {
+	hashValueBase := math.MaxUint64 / uint64(b.N)
+	deckSize, handSize := int32(512), int32(8)
+	pick := func(int32) {}
+	for i := 0; i < b.N; i++ {
+		ShuffleAndDeal(hashValueBase*uint64(i), deckSize, handSize, pick)
+	}
+}
+
+func TestShuffleAndDealToSliceWithValidation(t *testing.T) {
+	tests := []struct {
+		name      string
+		deckSize  int32
+		handSize  int32
+		validated bool
+	}{
+		{
+			"validation fails",
+			-100,
+			-100,
+			false,
+		},
+		{
+			"deckSize = handSize = 4",
+			4,
+			4,
+			true,
+		},
+		{
+			"deckSize = handSize = 8",
+			8,
+			8,
+			true,
+		},
+		{
+			"deckSize = handSize = 10",
+			10,
+			10,
+			true,
+		},
+		{
+			"deckSize = handSize = 12",
+			12,
+			12,
+			true,
+		},
+		{
+			"deckSize = 128, handSize = 8",
+			128,
+			8,
+			true,
+		},
+		{
+			"deckSize = 256, handSize = 7",
+			256,
+			7,
+			true,
+		},
+		{
+			"deckSize = 512, handSize = 6",
+			512,
+			6,
+			true,
+		},
+	}
+	for _, test := range tests {
+		hashValue := rand.Uint64()
+		t.Run(test.name, func(t *testing.T) {
+			cards, err := ShuffleAndDealToSliceWithValidation(hashValue, test.deckSize, test.handSize)
+			if (err == nil) != test.validated {
+				t.Errorf("test case %s fails in validation check", test.name)
+				return
+			}
+
+			if test.validated {
+				// check cards number
+				if len(cards) != int(test.handSize) {
+					t.Errorf("test case %s fails in cards number", test.name)
+					return
+				}
+
+				// check cards range and duplication
+				cardMap := make(map[int32]struct{})
+				for _, cardIdx := range cards {
+					if cardIdx < 0 || cardIdx >= test.deckSize {
+						t.Errorf("test case %s fails in range check", test.name)
+						return
+					}
+					cardMap[cardIdx] = struct{}{}
+				}
+				if len(cardMap) != int(test.handSize) {
+					t.Errorf("test case %s fails in duplication check", test.name)
+					return
+				}
+			}
+		})
+	}
+}
+
+func BenchmarkShuffleAndDealToSlice(b *testing.B) {
+	hashValueBase := math.MaxUint64 / uint64(b.N)
+	deckSize, handSize := int32(512), int32(8)
+	for i := 0; i < b.N; i++ {
+		_ = ShuffleAndDealToSlice(hashValueBase*uint64(i), deckSize, handSize)
+	}
+}
+
+func TestUniformDistribution(t *testing.T) {
+	deckSize, handSize := int32(128), int32(3)
+	handCoordinateMap := make(map[int]int)
+
+	allCoordinateCount := 128 * 127 * 126 / 6
+
+	for i := 0; i < allCoordinateCount*16; i++ {
```

Comment:
Even more interesting would be a power of 2 (the lowest one that is big enough).